overlord/snapstate/handlers.go

```diff
@@ -1147,6 +1149,27 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) error {
 		}
 		return copyDataErr
 	}
+
+	// the migration hasn't been done - do it now
+	if opts.UseHidden && !opts.MigratedToHidden {
+		err = m.backend.HideSnapData(snapsup.InstanceName())
+		snapsup.MigratedHidden = true
```

Comment:
I missed that in the previous state of this code, we still set MigradeHidden = true, because the migration may have suceeded for some users? If so, it's probably worth leaving a comment about that.

As a separate problem, looking at the backend HideSnapData(), it stops on the first error, so it is possible that some users will be migrated after all, but undo will not run, since the whole task handler would fail. Next, when the previous revision of snap runs, it would see ~/snap for some users, and ~/.snap for others. Looking at cmd_run code, it seems that if the feature is on we just assume that the new path is ~/.snap regardless of the user not being migrated?

Reply:
>  missed that in the previous state of this code, we still set MigradeHidden = true, because the migration may have suceeded for some users? If so, it's probably worth leaving a comment about that.

Yes, that's what the comment on the next line is about, I moved it down with the refactor but I can pull it back up.

> As a separate problem, looking at the backend HideSnapData(), it stops on the first error, so it is possible that some users will be migrated after all, but undo will not run, since the whole task handler would fail.

I see your point, I thought that the undo handler was called if this handler failed. If not then, I'll just add that call explicitly. Thanks for pointing that out.

> Looking at cmd_run code, it seems that if the feature is on we just assume that the new path is ~/.snap regardless of the user not being migrated?
The cmd_run.go code isn't updated in this PR because I had to cut it out make it shorter but it will read the sequence file to find out the status of the migration. I left a note in the PR description as well as a couple of TODOs in the /cmd code where the changes will be required.